app/repositories/UserDataRepository.scala

```diff
@@ -0,0 +1,139 @@
+import org.mongodb.scala.model.{FindOneAndReplaceOptions, FindOneAndUpdateOptions}
+import play.api.libs.json.Format
+import uk.gov.hmrc.mongo.play.json.Codecs.{logger, toBson}
+import uk.gov.hmrc.mongo.play.json.PlayMongoRepository
+import uk.gov.hmrc.mongo.play.json.formats.MongoJavatimeFormats
+import utils.PagerDutyHelper.PagerDutyKeys.{ENCRYPTION_DECRYPTION_ERROR, FAILED_TO_CREATE_DATA, FAILED_TO_FIND_DATA, FAILED_TO_UPDATE_DATA}
+import utils.PagerDutyHelper.pagerDutyLog
+
+import java.time.{Clock, Instant}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+trait UserDataRepository[C <: UserDataTemplate] {
+  self: PlayMongoRepository[C] =>
+  implicit val ec: ExecutionContext
+  implicit val instantFormat: Format[Instant] = MongoJavatimeFormats.instantFormat
+
+  val repoName: String
+  type UserData
+
+  def encryptionMethod: UserData => C
+
+  def decryptionMethod: C => UserData
+
+  def create[T](userData: UserData): Future[Either[DatabaseError, Unit]] = {
+    lazy val start = s"[$repoName][create]"
+    Try {
+      encryptionMethod(userData)
+    }.toOption match {
+      case Some(exception: Exception) => Future.successful(handleEncryptionDecryptionException(exception, start))
+      case Some(encryptedData) =>
+        collection.insertOne(encryptedData).toFutureOption().map {
+          case Some(_) => Right(())
+          case None => Left(DataNotUpdated)
+        }.recover {
+          case exception: Exception =>
+            pagerDutyLog(FAILED_TO_CREATE_DATA, s"$start Failed to create user data. Exception: ${exception.getMessage}")
+            Left(DataNotUpdated)
+        }
+      case _ => Future(Left(DataNotUpdated))
+    }
+  }
+
+  def find[T](taxYear: Int)(implicit user: User[_]): Future[Either[DatabaseError, Option[UserData]]] = {
+    lazy val start = s"[$repoName][find]"
+
+    val userData = collection.findOneAndUpdate(
+      filter = filter(user.sessionId, user.mtditid, user.nino, taxYear),
+      update = set("lastUpdated", toBson(Instant.now(Clock.systemUTC()))),
+      options = FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER)
+    ).toFutureOption().map {
+      case Some(data) => Right(Some(data))
+      case None =>
+        logger.info(s"$start No CYA data found for user. SessionId: ${user.sessionId}")
+        Right(None)
+    }.recover {
+      case exception: Exception =>
+        pagerDutyLog(FAILED_TO_FIND_DATA, s"$start Failed to find user data. Exception: ${exception.getMessage}")
+        Left(DataNotFound)
+    }
+
+    userData.map {
+      case Left(_) => Left(DataNotFound)
+      case Right(data) =>
+        Try {
+          data.map(decryptionMethod)
+        }.toEither match {
+          case Left(value) => handleEncryptionDecryptionException(value.asInstanceOf[Exception], start)
+          case Right(value: Option[UserData]) => Right(value)
+        }
+    }
+
+  }
+
+  def update(userData: UserData): Future[Either[DatabaseError, Boolean]] = {
```

Comment:
Can be changed to `Future[Either[DatabaseError, Unit]]` and return `Right(())`

Reply:
Thanks, will update in the next commit.